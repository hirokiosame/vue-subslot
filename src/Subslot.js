import { emit } from './utils/emit';
import { filterVnodes } from './utils/filter-vnodes';
import createDefineMixin from './create-define-mixin';

export default {
	functional: true,
	props: {
		not: {
			type: Boolean,
		},
		element: {
			type: [Object, Array, String],
		},
		offset: {
			type: [String, Number],
			default: 0,
			validator: val => !Number.isNaN(parseInt(val, 10)),
		},
		limit: {
			type: [String, Number],
			validator: val => !Number.isNaN(parseInt(val, 10)),
		},
		vnodes: {
			type: null,
		},
		name: {
			type: String,
			default: 'default',
		},
	},

	render(h, ctx) {
		const { props, parent } = ctx;

		let vnodes;

		// Detect definition
		if (parent.$subslots) {
			vnodes = parent.$subslots[props.name];
		} else {
			vnodes = ctx.props.vnodes || ctx.parent.$slots.default || [];
			vnodes = filterVnodes({
				vnodes,
				filter: props,
				vm: parent,
			});
		}

		if (!vnodes || vnodes.length === 0) {
			emit(ctx, 'no-match');
			return ctx.slots().default;
		}

		return vnodes;
	},

	/* Static method for mixin */
	define: createDefineMixin,
};
