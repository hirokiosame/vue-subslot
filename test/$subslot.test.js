import { mount } from '@vue/test-utils';
import Subslot from 'vue-subslot';
import CardHeader from './fixtures/CardHeader.vue';
import CardFooter from './fixtures/CardFooter.vue';

describe('$subslots support', () => {

	test('Normal usage', () => {
		const Card = {
			template: `
				<div class="card">
					<div
						v-if="$subslots.cardHeader"
						class="card-header"
					>
						<subslot name="cardHeader" />
					</div>

					<div class="card-content">
						<subslot />
					</div>

					<div class="card-footer">
						<subslot name="cardFooter" />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
					cardFooter: {
						element: CardFooter,
						limit: 1,
					},
				}),
			],
		};

		const usage = {
			template: `
				<card>
					<card-header>
						Header
					</card-header>

					Content

					<card-footer>
						Footer
					</card-footer>
				</card>
			`,
			components: {
				Card,
				CardHeader,
				CardFooter,
			},
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Empty slot', () => {
		const Card = {
			template: `
				<div class="card">
					<div
						v-if="$subslots.cardHeader"
						class="card-header"
					>
						<subslot name="cardHeader" />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
				}),
			],
		};

		const usage = {
			template: '<card>Content</card>',
			components: { Card },
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should use fallback', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot name="cardHeader">
							Fallback
						</subslot>
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
				}),
			],
		};

		const usage = {
			template: '<card>Content</card>',
			components: { Card },
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Reactive subslots', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot name="cardHeader" />
					</div>
					<div class="card-content">
						<subslot />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
				}),
			],
		};

		const usage = {
			template: `
				<card>
					<card-header>
						Header {{ msg }}
					</card-header>

					Content {{ msg }}
				</card>
			`,

			components: {
				Card,
				CardHeader,
			},

			data() {
				return { msg: 'is not reactive' };
			},

			methods: {
				increment() {
					this.msg = 'is reactive!';
				},
			},
		};

		const wrapper = mount(usage);
		wrapper.vm.increment();
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Fallback reactive', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot name="cardHeader">
						Fallback {{ msg }}
					</subslot>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
				}),
			],

			data() {
				return { msg: 'is not reactive' };
			},

			mounted() {
				this.msg = 'is reactive!';
			},
		};

		const usage = {
			template: `
			<card>
				Content
			</card>
			`,
			components: {
				Card,
				CardHeader,
			},
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support slice index', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot name="cardHeader" />
					</div>

					<div class="card-content">
						<subslot />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader[2]',
				}),
			],
		};

		const usage = {
			template: `
				<card>
					<card-header
						v-for="i in 10"
						:key="i"
					>
						Header {{ i }}
					</card-header>
				</card>
			`,
			components: {
				Card,
				CardHeader,
				CardFooter,
			},
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support slice with limit', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot name="cardHeader" />
					</div>

					<div class="card-content">
						<subslot />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader[2:2]',
				}),
			],
		};

		const usage = {
			template: `
				<card>
					<card-header
						v-for="i in 10"
						:key="i"
					>
						Header {{ i }}
					</card-header>
				</card>
			`,
			components: {
				Card,
				CardHeader,
				CardFooter,
			},
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support slice with no offset', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot name="cardHeader" />
					</div>

					<div class="card-content">
						<subslot />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader[:2]', // Equivalent to [0:2]
				}),
			],
		};

		const usage = {
			template: `
				<card>
					<card-header
						v-for="i in 10"
						:key="i"
					>
						Header {{ i }}
					</card-header>
				</card>
			`,
			components: {
				Card,
				CardHeader,
				CardFooter,
			},
		};

		const wrapper = mount(usage);
		expect(wrapper.element).toMatchSnapshot();
	});
	test('Should emit "no-match" on no match', () => {
		const onNoCardHeader = jest.fn();

		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot
							name="cardHeader"
							@no-match="onNoCardHeader"
						/>
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},

			mixins: [
				Subslot.define({
					cardHeader: '@CardHeader:1',
				}),
			],

			methods: {
				onNoCardHeader,
			},
		};

		const usage = {
			template: '<card>Content</card>',
			components: { Card },
		};

		const wrapper = mount(usage);
		expect(onNoCardHeader).toBeCalled();
		expect(wrapper.element).toMatchSnapshot();
	});
});
