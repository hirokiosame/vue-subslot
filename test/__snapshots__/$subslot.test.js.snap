// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$subslots support Empty slot 1`] = `
<div
  class="card"
>
  <!---->
</div>
`;

exports[`$subslots support Fallback reactive 1`] = `
<div
  class="card"
>
  
						Fallback is reactive!
					
</div>
`;

exports[`$subslots support Normal usage 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    

					Content

					
  </div>
   
  <div
    class="card-footer"
  >
    <div>
      
						Footer
					
    </div>
  </div>
</div>
`;

exports[`$subslots support Reactive subslots 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header is reactive!
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    

					Content is reactive!
				
  </div>
</div>
`;

exports[`$subslots support Should emit "no-match" on no match 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <!---->
  </div>
</div>
`;

exports[`$subslots support Should support slice index 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header 3
					
    </div>
    <div>
      
						Header 4
					
    </div>
    <div>
      
						Header 5
					
    </div>
    <div>
      
						Header 6
					
    </div>
    <div>
      
						Header 7
					
    </div>
    <div>
      
						Header 8
					
    </div>
    <div>
      
						Header 9
					
    </div>
    <div>
      
						Header 10
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    <div>
      
						Header 1
					
    </div>
    <div>
      
						Header 2
					
    </div>
  </div>
</div>
`;

exports[`$subslots support Should support slice with limit 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header 3
					
    </div>
    <div>
      
						Header 4
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    <div>
      
						Header 1
					
    </div>
    <div>
      
						Header 2
					
    </div>
    <div>
      
						Header 5
					
    </div>
    <div>
      
						Header 6
					
    </div>
    <div>
      
						Header 7
					
    </div>
    <div>
      
						Header 8
					
    </div>
    <div>
      
						Header 9
					
    </div>
    <div>
      
						Header 10
					
    </div>
  </div>
</div>
`;

exports[`$subslots support Should support slice with no offset 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header 1
					
    </div>
    <div>
      
						Header 2
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    <div>
      
						Header 3
					
    </div>
    <div>
      
						Header 4
					
    </div>
    <div>
      
						Header 5
					
    </div>
    <div>
      
						Header 6
					
    </div>
    <div>
      
						Header 7
					
    </div>
    <div>
      
						Header 8
					
    </div>
    <div>
      
						Header 9
					
    </div>
    <div>
      
						Header 10
					
    </div>
  </div>
</div>
`;

exports[`$subslots support Should use fallback 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    
							Fallback
						
  </div>
</div>
`;
